Enhance directory traversal handling/resolution.

barryhughes opened the pull request:
### Changes proposed in this Pull Request:

This set of changes is a necessary precursor to some other changes that are required to support shortcodes as a means of supplying a downloadable file URL (as is the case with the S3 downloads add-on).

Specifically, it guards against tricky issues that might occur if a plugin like S3 is disabled and/or other scenarios where a disused shortcode is supplied to represent a downloadable filepath or URL (because in those cases, the shortcode is  unparseable and essentially represents a relative filepath.

Additionally, it improves logic relating to handling of filepaths and URLs that contain directory traversals from within our internally-facing URL utility class.

### How to test the changes in this Pull Request:

There is no UI-driven way to assess this set of changes. Instead, please review the newly added tests, and revisions to the existing tests. Those are probably the best indication of the scenarios I wanted to touch on.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

Not needed.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

barryhughes committed:
Enhance directory traversal handling/resolution.

barryhughes committed:
Cover further scenarios relating to relative URLs and traversals.

barryhughes committed:
Tweak commentary.

barryhughes committed:
Improve test description.

vedanshujain commented on plugins/woocommerce/src/Internal/Utilities/URL.php:
```diff
@@ -124,8 +131,11 @@ private function preprocess() {
-		$segments          = explode( '/', $this->components['path'] );
-		$this->is_absolute = substr( $this->components['path'], 0, 1 ) === '/';
+		$segments           = explode( '/', $this->components['path'] );
+		$this->is_absolute  = substr( $this->components['path'], 0, 1 ) === '/' || ! empty( $this->components['host'] );
```
does `absolute` here mean pointing to a specific file (as opposed to a relative URL or file)? if so, then what would be the case where `! empty( $this->components['host'] )` would also be needed (just curious)?

barryhughes replied:
Yes, in the case of file URLs we are checking if it represents an absolute filepath, whether that is a directory or a file. 

For URLs with host components (generally, anything that is not a file URL), we also want to treat them as being absolute specifically in relation to how any leading directory traversals are handled.

The inputs and outputs we want (simplified, but covers this aspect):

| Type | Unprocessed (Host+)Path | Processed Path |
| --- | --- | --- |
| **Absolute** | `/../../remainder` | `/remainder` |
| **Absolute** | `example.com/../../remainder` | `example.com/remainder` |
| **Relative** | `../../remainder` | `../../remainder` |

vedanshujain commented on plugins/woocommerce/src/Internal/Utilities/URL.php:
```diff
@@ -124,8 +131,11 @@ private function preprocess() {
-		$this->is_absolute = substr( $this->components['path'], 0, 1 ) === '/';
+		$segments           = explode( '/', $this->components['path'] );
+		$this->is_absolute  = substr( $this->components['path'], 0, 1 ) === '/' || ! empty( $this->components['host'] );
+		$this->is_directory = substr( $this->components['path'], -1, 1 ) === '/' && strlen( $this->components['path'] ) > 1;
```
why `strlen( $this->components['path'] ) > 1;`? just `/` is also a directory right?

barryhughes replied:
Yes, it is also a directory; I'll think about a different name as `$is_directory` is perhaps a little misleading (as to why we do this, one of the reasons is that when we reconstitute the path at the end of processing, we don't want to apply an extra trailing slash to root dir `/` (ie, we don't want `//`).

vedanshujain commented on plugins/woocommerce/src/Internal/Utilities/URL.php:
```diff
@@ -137,19 +147,43 @@ private function process_path() {
-				continue;
+			// Resolve directory traversals (if allowed: see further comment relating to this).
+			if ( $resolve_traversals && $is_traversal ) {
+				if ( count( $this->path_parts ) > 0 && ! $retain_traversals ) {
```
there might be an edge case when the path starts from `./`, for example, `./.././../foo.txt`

vedanshujain replied:
what do we do when the path starts from other special chars, such as `~`, `-` etc?

barryhughes replied:
Both are treated as any other character (and, though `~` has special meaning in filepaths, I don't believe `-` does?).

I'll look at documenting what we expect for those cases (and the possible edge case) via further unit tests, though, along with the others you suggest further down :+1:

barryhughes replied:
This note doesn't necessarily require a follow-up, I just thought there was a chance it may be handy to record a note someplace: though of course `~` at the start of a path is expanded by most shells to the current user home dir, it is not treated that way when working with `fopen()` and related functions: in those cases, it is just another symbol (for instance, if the current working dir is `/tmp`, then `~/file.txt` would mean `/tmp/~/file.txt` ... an unusual but valid path).

vedanshujain commented on plugins/woocommerce/tests/php/src/Internal/Utilities/URLTest.php:
```diff
@@ -42,7 +42,19 @@ public function test_directory_traversal_resolution() {
+			'Directory traversals are appropriately resolved even in complex cases with multiple separate traversals. When the original path is absolute, the output will be absolute.'
 		);
 	}
 
```
Let's add more tests:
`./.././.././../foo.txt`
`~/foo.txt`
`///foo`
`bazz//foo`
`foo://bar`
`http://localhost?../../bar`

vedanshujain reviewed (commented):
Added some preliminary comments, but still testing in 32362

barryhughes committed:
Rename the is_directory property to give a better sense of its role.

barryhughes committed:
Improve path resolution, add additional tests, clean-up tests (use data-providers).

barryhughes commented:
Updated (and rebased following PHP 7.2-related changes that were merged to trunk).

@vedanshujain can you take another look? I also switched to using data providers in the tests, I felt this was more readable and it makes it fast and easy to drop-in additional scenarios if needed.

vedanshujain reviewed (approved):
Looks pretty neat!

barryhughes commented:
(Changelog/further testing instructions not required: this is an amendment to an as-yet-unreleased change.)
